### DISCUSSION TITLE
Documentation with guidance on how often to upgrade ZBM and how

### DISCUSSION THREAD

**Discussion by timkgh** [zbm-dev/zfsbootmenu#575] [Feature Requests]
Any guidance on how often to update ZBM and how to do it, e.g. running on Ubuntu 22.04?
Any consideration for a PPA that would make this management easier?

**Comment by zdykstra** [comment]
Upgrading ZFSBootMenu depends on the method of installation. When you do it largely depends on two factors:

1) Are there new features or bug fixes that I need?
2) Is the version of ZFS in ZFSBootMenu now too old for my system?

You're likely fine for some time - Ubuntu 22.04 doesn't ship a very recent version of ZFS, so there shouldn't be any pool compatibility issues.

We won't be providing a PPA or any other distro-specific packaging infrastructure.

**Reply by timkgh** [reply]
FWIW I want to thank you for writing the [portable ZBM](https://docs.zfsbootmenu.org/en/v2.3.x/guides/general/portable.html) instructions! ❤️ 
It helped me boot from ZFS again, I just ran into this grub issue:
https://github.com/openzfs/zfs/issues/13873#issuecomment-1889885090

Now I'm trying to figure out what to do with `/boot`. Move it to its own bpool (which IIUC would prevent me from using ZBM), switch to ZBM, move /boot to ext4 and avoid future grub issues.

**Reply by zdykstra** [reply]
Using a dedicated `/boot` dataset (either on the same pool, or a different one) would prevent ZFSBootMenu from booting your system; it needs `/boot` to be a sub-directory of a dataset that mounts to `/`. 

We're obviously biased, but you'd be fine running ZFSBootMenu with Ubuntu. It needs fairly infrequent updates and worst case, as you've noted, you can just throw a copy on to a USB drive and boot your system.

**Comment by timkgh** [comment]
> Ubuntu 22.04 doesn't ship a very recent version of ZFS, so there shouldn't be any pool compatibility issues

Any guidance for the upcoming 24.04 which I'll probably move to later this year?

**Reply by classabbyamp** [reply]
[appears to be 2.2.2](https://packages.ubuntu.com/noble/zfs-dkms), which would work with the current prebuilt images.

**Reply by Piotr2009** [reply]
hello) thanks for a great work !  installed on debian 12.2 bookworm . lost some time and efforts when following from the start the tutorial of openzfs for debian , dual pools bootfs/rootfs is not the case here. Question - should i expect problems when upgrading debian to trixie , thats version 13 . I have compiled from source zfsbootmenu . Is it supported or better not-go ? . thanks

**Reply by zdykstra** [reply]
I haven't tried building ZFSBootMenu from source on Debian 13, so I can't explicitly say that it will work. However, I would expect that it'll continue to work with out any real issues. If you do run into problems with it, feel free to create a new discussion and we can possibly help you work through the problem.

**Reply by Piotr2009** [reply]
Super !  the upgrade went flawlessly . i'm very happy. debian bookworm -> trixie  Update , upgrade, plus zbm-generate  and viola . чудово)

**Comment by timkgh** [comment]
Any tutorial with clear steps on how to replace GRUB with ZBM on Ubuntu? Or maybe not _replace_ but rather have the option to boot from either, if one is broken, there's the other one as a backup.
I have a high level idea on how things work, but I'm interested in the details, how to properly configure it, auto boot an entry (after some timeout that lets me select a different kernel), what happens when Ubuntu updates the kernel, do I need to manually do something with ZBM to be able to boot the previous kernel version if the new one has issues, etc.

**Reply by zdykstra** [reply]
If your system is already booting from ZFSBootMenu, then the basics are already in place for you. Configuration is done in two key ways - pool/dataset properties and the ZFSBootMenu kernel command line. Both are documented at https://docs.zfsbootmenu.org/en/v2.3.x/man/zfsbootmenu.7.html . 

In short, set `bootfs` on the pool to control the default boot entry. Set `org.zfsbootmenu:commandline` on the dataset to control the commandline passed into the kernel.  Previous kernels can be selected/booted/pinned through the interface or through a dataset property.

**Reply by timkgh** [reply]
I need to figure out what kernel params I may need to pass, whatever grub used to pass, if anything, though just booting from the portable ZBM seems to work just fine, so perhaps just stay with the simplest defaults, nothing to configure.

**Reply by timkgh** [reply]
Posting here, hopefully others discover it too and also to check my work ...

It appears `grub` passes the following kernel params (from `/boot/grub/grub.cfg` menu entries because I cannot boot it):
`linux   "/ROOT/ubuntu@/boot/vmlinuz-6.5.0-14-generic" root=ZFS="rpool/ROOT/ubuntu" ro fbcon=font:TER16x32 crashkernel=512M-:192M`

ZBM passed these params (from `cat /proc/cmdline` of a live system booted with ZBM defaults):
`root=zfs:rpool/ROOT/ubuntu quiet loglevel=4 spl.spl_hostid=0x...`

The format of `root=` is different but looks reasonable and perhaps alternate syntax, `ro` is the default anyway IIUC and the rest not sure they really matter in practice.

I'll remove `quiet loglevel=4` from ZBM because I like a verbose boot to notice errors.

Not sure what value the `bootfs` property should have. My rpool doesn't have it set at all. Does it matter to set it? I have another zpool on the system, but it's just storage it doesn't have any kernels, etc.
Edit: found it [here](https://openzfs.github.io/openzfs-docs/man/v2.2/7/zpoolprops.7.html):
```
bootfs=(unset)|pool[/dataset]
Identifies the default bootable dataset for the root pool. This property is expected to be set mainly by the installation and upgrade programs. Not all Linux distribution boot processes use the bootfs property.
```
Looks like I need to set `bootfs=rpool/ROOT/ubuntu`

**Reply by classabbyamp** [reply]
`root` and `spl.spl_hostid` are added by zbm when booting so they should not be included when setting the dataset property

**Reply by zdykstra** [reply]
You can set bootfs in the ZBM interface via MOD+D, otherwise you can do it via `zpool set ...`. Once that's set, that dataset will automatically boot with out any required user interaction.

**Reply by timkgh** [reply]
I got it all working to my liking. Amazing piece of software! Thank you so much for all your work. ❤️
